gps/prune.go

```diff
@@ -41,24 +31,85 @@ const (
 	PruneGoTestFiles
 )
 
-// DefaultRootPruneOptions instantiates a copy of the default root prune options.
-func DefaultRootPruneOptions() RootPruneOptions {
-	return RootPruneOptions{
-		PruneOptions:   PruneNestedVendorDirs,
-		ProjectOptions: PruneProjectOptions{},
-	}
+// PruneOptionSet represents trinary distinctions for each of the types of
+// prune rules (as expressed via PruneOptions): nested vendor directories,
+// unused packages, non-go files, and go test files.
+//
+// The three-way distinction is between "none", "true", and "false", represented
+// by uint8 values of 0, 1, and 2, respectively.
+//
+// This trinary distinction is necessary in order to record, with full fidelity,
+// a cascading tree of pruning values, as expressed in CascadingPruneOptions; a
+// simple boolean cannot delineate between "false" and "none".
+type PruneOptionSet struct {
+	NestedVendor   uint8
+	UnusedPackages uint8
+	NonGoFiles     uint8
+	GoTests        uint8
+}
+
+// CascadingPruneOptions is a set of rules for pruning a dependency tree.
+//
+// The DefaultOptions are the global default pruning rules, expressed as a
+// single PruneOptions bitfield. These global rules will cascade down to
+// individual project rules, unless superseded.
+type CascadingPruneOptions struct {
+	DefaultOptions    PruneOptions
+	PerProjectOptions map[ProjectRoot]PruneOptionSet
 }
 
-// PruneOptionsFor returns the prune options for the passed project root.
+// PruneOptionsFor returns the PruneOptions bits for the given project,
+// indicating which pruning rules should be applied to the project's code.
 //
-// It will return the root prune options if the project does not have specific
-// options or if it does not exist in the manifest.
-func (o *RootPruneOptions) PruneOptionsFor(pr ProjectRoot) PruneOptions {
-	if po, ok := o.ProjectOptions[pr]; ok {
-		return po
+// It computes the cascade from default to project-specific options (if any) on
+// the fly.
+func (o CascadingPruneOptions) PruneOptionsFor(pr ProjectRoot) PruneOptions {
+	po, has := o.PerProjectOptions[pr]
+	if !has {
+		return o.DefaultOptions
+	}
+
+	ops := o.DefaultOptions
+	if po.NestedVendor != 0 {
+		if po.NestedVendor == 1 {
+			ops |= PruneNestedVendorDirs
+		} else {
+			ops &^= PruneNestedVendorDirs
+		}
 	}
 
-	return o.PruneOptions
+	if po.UnusedPackages != 0 {
+		if po.UnusedPackages == 1 {
+			ops |= PruneUnusedPackages
+		} else {
+			ops &^= PruneUnusedPackages
+		}
+	}
+
+	if po.NonGoFiles != 0 {
+		if po.NonGoFiles == 1 {
+			ops |= PruneNonGoFiles
+		} else {
+			ops &^= PruneNonGoFiles
+		}
+	}
+
+	if po.GoTests != 0 {
+		if po.GoTests == 1 {
+			ops |= PruneGoTestFiles
+		} else {
+			ops &^= PruneGoTestFiles
+		}
+	}
+
+	return ops
+}
+
+func defaultCascadingPruneOptions() CascadingPruneOptions {
```

Comment:
Maybe export this instead of having multiple copies of it?

Reply:
i intentionally didn't export it, as i don't think it's worth polluting the package interface with it.